Tensor: Define operators override functions in TensorBody.h

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #69840
* #68462
* #68693
* #68692
* #68691
* __->__ #68697
* #68690
* #68689
* #68714
* #68688
* #68687
* #68247

Currently, if you include `Tensor.h` but not `TensorOperators.h` then
using overloaded operators will compile but fail at link time.
Instead, this defines the member functions in `TensorBody.h` and
leaves `TensorOperators.h` as only the free functions.

Differential Revision: [D32596269](https://our.internmc.facebook.com/intern/diff/D32596269)

peterbell10 committed:
Tensor: Define operators override functions in TensorBody.h

Currently, if you include `Tensor.h` but not `TensorOperators.h` then
using overloaded operators will compile but fail at link time.
Instead, this defines the member functions in `TensorBody.h` and
leaves `TensorOperators.h` as only the free functions.

[ghstack-poisoned]

peterbell10 committed:
Update on "Tensor: Define operators override functions in TensorBody.h"


Currently, if you include `Tensor.h` but not `TensorOperators.h` then
using overloaded operators will compile but fail at link time.
Instead, this defines the member functions in `TensorBody.h` and
leaves `TensorOperators.h` as only the free functions.

[ghstack-poisoned]

peterbell10 committed:
Update on "Tensor: Define operators override functions in TensorBody.h"


Currently, if you include `Tensor.h` but not `TensorOperators.h` then
using overloaded operators will compile but fail at link time.
Instead, this defines the member functions in `TensorBody.h` and
leaves `TensorOperators.h` as only the free functions.

[ghstack-poisoned]

peterbell10 committed:
Update on "Tensor: Define operators override functions in TensorBody.h"


Currently, if you include `Tensor.h` but not `TensorOperators.h` then
using overloaded operators will compile but fail at link time.
Instead, this defines the member functions in `TensorBody.h` and
leaves `TensorOperators.h` as only the free functions.

[ghstack-poisoned]

albanD commented on aten/src/ATen/templates/TensorBody.h:
```diff
@@ -242,22 +248,67 @@ class TORCH_API Tensor: public TensorBase {
+    if (!index.isIntegral(false)) {
+      TORCH_CHECK_INDEX(false, "Can only index tensors with integral scalars");
+    }
+    return select(0, index.toLong());
```
nit: I know it was here before, but to be consistent with the one below, this could do `return this->operator[](index.toLong())`;

albanD reviewed (approved):
SGTM

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596269).

peterbell10 committed:
Update on "Tensor: Define operators override functions in TensorBody.h"


Currently, if you include `Tensor.h` but not `TensorOperators.h` then
using overloaded operators will compile but fail at link time.
Instead, this defines the member functions in `TensorBody.h` and
leaves `TensorOperators.h` as only the free functions.

Differential Revision: [D32596269](https://our.internmc.facebook.com/intern/diff/D32596269)

[ghstack-poisoned]

peterbell10 committed:
Update on "Tensor: Define operators override functions in TensorBody.h"


Currently, if you include `Tensor.h` but not `TensorOperators.h` then
using overloaded operators will compile but fail at link time.
Instead, this defines the member functions in `TensorBody.h` and
leaves `TensorOperators.h` as only the free functions.

Differential Revision: [D32596269](https://our.internmc.facebook.com/intern/diff/D32596269)

[ghstack-poisoned]

peterbell10 committed:
Update on "Tensor: Define operators override functions in TensorBody.h"


Currently, if you include `Tensor.h` but not `TensorOperators.h` then
using overloaded operators will compile but fail at link time.
Instead, this defines the member functions in `TensorBody.h` and
leaves `TensorOperators.h` as only the free functions.

Differential Revision: [D32596269](https://our.internmc.facebook.com/intern/diff/D32596269)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596269).

peterbell10 committed:
Rebase to fix merge conflicts on "Tensor: Define operators override functions in TensorBody.h"


Currently, if you include `Tensor.h` but not `TensorOperators.h` then
using overloaded operators will compile but fail at link time.
Instead, this defines the member functions in `TensorBody.h` and
leaves `TensorOperators.h` as only the free functions.

Differential Revision: [D32596269](https://our.internmc.facebook.com/intern/diff/D32596269)

[ghstack-poisoned]

peterbell10 committed:
Update on "Tensor: Define operators override functions in TensorBody.h"


Currently, if you include `Tensor.h` but not `TensorOperators.h` then
using overloaded operators will compile but fail at link time.
Instead, this defines the member functions in `TensorBody.h` and
leaves `TensorOperators.h` as only the free functions.

Differential Revision: [D32596269](https://our.internmc.facebook.com/intern/diff/D32596269)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596269).

peterbell10 committed:
Update on "Tensor: Define operators override functions in TensorBody.h"


Currently, if you include `Tensor.h` but not `TensorOperators.h` then
using overloaded operators will compile but fail at link time.
Instead, this defines the member functions in `TensorBody.h` and
leaves `TensorOperators.h` as only the free functions.

Differential Revision: [D32596269](https://our.internmc.facebook.com/intern/diff/D32596269)

[ghstack-poisoned]

peterbell10 committed:
Fix mypy issues on "Tensor: Define operators override functions in TensorBody.h"


Currently, if you include `Tensor.h` but not `TensorOperators.h` then
using overloaded operators will compile but fail at link time.
Instead, this defines the member functions in `TensorBody.h` and
leaves `TensorOperators.h` as only the free functions.

Differential Revision: [D32596269](https://our.internmc.facebook.com/intern/diff/D32596269)

[ghstack-poisoned]

peterbell10 committed:
Update on "Tensor: Define operators override functions in TensorBody.h"


Currently, if you include `Tensor.h` but not `TensorOperators.h` then
using overloaded operators will compile but fail at link time.
Instead, this defines the member functions in `TensorBody.h` and
leaves `TensorOperators.h` as only the free functions.

Differential Revision: [D32596269](https://our.internmc.facebook.com/intern/diff/D32596269)

[ghstack-poisoned]

peterbell10 committed:
Update on "Tensor: Define operators override functions in TensorBody.h"


Currently, if you include `Tensor.h` but not `TensorOperators.h` then
using overloaded operators will compile but fail at link time.
Instead, this defines the member functions in `TensorBody.h` and
leaves `TensorOperators.h` as only the free functions.

Differential Revision: [D32596269](https://our.internmc.facebook.com/intern/diff/D32596269)

[ghstack-poisoned]

peterbell10 committed:
Commit to re-run CI on "Tensor: Define operators override functions in TensorBody.h"


Currently, if you include `Tensor.h` but not `TensorOperators.h` then
using overloaded operators will compile but fail at link time.
Instead, this defines the member functions in `TensorBody.h` and
leaves `TensorOperators.h` as only the free functions.

Differential Revision: [D32596269](https://our.internmc.facebook.com/intern/diff/D32596269)

[ghstack-poisoned]

peterbell10 committed:
Update on "Tensor: Define operators override functions in TensorBody.h"


Currently, if you include `Tensor.h` but not `TensorOperators.h` then
using overloaded operators will compile but fail at link time.
Instead, this defines the member functions in `TensorBody.h` and
leaves `TensorOperators.h` as only the free functions.

Differential Revision: [D32596269](https://our.internmc.facebook.com/intern/diff/D32596269)

[ghstack-poisoned]

peterbell10 committed:
Update on "Tensor: Define operators override functions in TensorBody.h"


Currently, if you include `Tensor.h` but not `TensorOperators.h` then
using overloaded operators will compile but fail at link time.
Instead, this defines the member functions in `TensorBody.h` and
leaves `TensorOperators.h` as only the free functions.

Differential Revision: [D32596269](https://our.internmc.facebook.com/intern/diff/D32596269)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596269).
